test/integration/master/kms_transformation_test.go

```diff
@@ -140,22 +123,80 @@ func TestKMSProvider(t *testing.T) {
 	}
 
 	// Secrets should be un-enveloped on direct reads from Kube API Server.
-	s, err := test.restClient.CoreV1().Secrets(testNamespace).Get(testSecret, metav1.GetOptions{})
+	mustReadSecret(t, test, true)
+	test.printMetrics()
+}
+
+// TestKMSFailureImpactOnReads validates that KMS Plugin will not be called for secrets that were cached by api-server\pkg\storage\cacher.go
+// Cacher.go (assuming sufficient memory) should intercept requests for secrets that were recently created or recently retrieved - thus
+// alleviating the need for caching within envelop.go.
+func TestKMSFailureImpactOnReads(t *testing.T) {
+	pluginMock, test := mustSetupTest(t)
+	defer pluginMock.cleanUp()
+	defer test.cleanUp()
+
+	pluginMock.grpcServer.GracefulStop()
+	// Giving time for mock plugin to shutdown
+	time.Sleep(2 * time.Second)
```

Comment:
2 seconds is not sufficient to avoid flakes when running in CI. Normally we'd use [`wait.ForeverTestTimeout`](https://github.com/kubernetes/kubernetes/blob/cdc411edf083ed89d1724e5af7bb5d7c7832a454/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L36), but you don't want to block the test for 30 seconds. Can you plumb a wait group or "done" channel through?